third_party/nvfuser/csrc/fusion_segmenter.cpp

```diff
@@ -3145,6 +3172,66 @@ void SegmentedFusion::setCachedHeuristicDataFor(
   heuristic_summary_cache_[group] = std::move(data);
 }
 
+void SegmentedFusion::validate(bool require_disjoint) const {
+  // Make sure the groups are a DAG
+  std::unordered_set<SegmentedGroup*> input_groups;
+  std::copy_if(
+      groups().begin(),
+      groups().end(),
+      std::inserter(input_groups, input_groups.end()),
+      [](SegmentedGroup* group) { return group->isInputGroup(); });
+
+  std::unordered_set<SegmentedEdge*> visited_edges;
+
+  std::deque<SegmentedGroup*> to_visit(
+      {input_groups.begin(), input_groups.end()});
+
+  while (!to_visit.empty()) {
```

Comment:
I'm not sure what the requirement is that's being checked here. I don't know what the sufficient check is for checking for a correct DAG. I guess I would attempt to traverse it and find if I get stuck in the traversal, but maybe the check you're trying here is simpler?

Reply:
There should have been `visited_edges.insert(consumer_edge)`.

Reply:
Don't you need to make sure all producer edges were visited before traversing a group? Otherwise I think you'd have to traverse starting at one input at a time no?

Reply:
I changed it to do DFS traversals in a similar way as `IterVisitor::traverseBetween`.